File: ts/indicators/john-ehlers/cyber-cycle/smoothing-factor-params.ts
import { BarComponent } from '../../../entities/bar-component';
import { QuoteComponent } from '../../../entities/quote-component';
import { TradeComponent } from '../../../entities/trade-component';

/** Describes parameters to create an instance of the CyberCycle indicator based on smoothing factor. */
export interface CyberCycleSmoothingFactorParams {
  /**
   * SmoothingFactor is the smoothing factor, α in [0,1], of the Cyber Cycle.
   *
   * The equivalent length ℓ is:
   *
   *     ℓ = round(2/α) - 1, 0<α≤1, 1≤ℓ.
   *
   * If α is near zero (< epsilon), ℓ is set to Number.MAX_SAFE_INTEGER.
   * The default value used by Ehlers is 0.07 (ℓ = 28).
   */
  smoothingFactor: number;

  /**
   * SignalLag is the signal lag (the number of time periods) for the signal line EMA.
   *
   * The signal EMA factor is 1/(signalLag+1).
   * The value should be a positive integer, greater or equal to 1.
   * The default value used by Ehlers is 9.
   */
  signalLag: number;

  /**
   * A component of a bar to use when updating the indicator with a bar sample.
   *
   * If _undefined_, the bar component defaults to BarComponent.Median (median price)
   * and will be shown in the indicator mnemonic as 'hl/2' since it differs from the
   * framework default (close).
   */
  barComponent?: BarComponent;

  /**
   * A component of a quote to use when updating the indicator with a quote sample.
   *
   * If _undefined_, the quote component will have a default value and will not be shown in the indicator mnemonic.
   */
  quoteComponent?: QuoteComponent;

  /**
   * A component of a trade to use when updating the indicator with a trade sample.
   *
   * If _undefined_, the trade component will have a default value and will not be shown in the indicator mnemonic.
   */
  tradeComponent?: TradeComponent;
}

File: go/indicators/johnehlers/cybercycle/smoothingfactorparams.go
package cybercycle

import "zpano/entities"

// SmoothingFactorParams describes parameters to create an instance of the Cyber Cycle indicator
// based on smoothing factor.
type SmoothingFactorParams struct {
	// SmoothingFactor is the smoothing factor, α, of the cyber cycle.
	// The equivalent length ℓ is:
	//
	//	ℓ = 2/α - 1, 0 < α ≤ 1, 1 ≤ ℓ
	//
	// The default value used in the Ehler's book is 0.07.
	SmoothingFactor float64

	// SignalLag is the lag, ℓ, of the signal line, which is an exponential moving average
	// of the cyber cycle value. The equivalent EMA smoothing factor α is:
	//
	//	α = 1/(ℓ + 1), 0 < α ≤ 1, 0 ≤ ℓ
	//
	// The value should be a positive integer, greater or equal to 1.
	// There are two default values used in the Ehler's book: 9 (EasyLanguage code) and 20 (EFS code).
	SignalLag int

	// BarComponent indicates the component of a bar to use when updating the indicator with a bar sample.
	//
	// If zero, the default (BarMedianPrice) is used and the component is not shown in the indicator mnemonic.
	BarComponent entities.BarComponent

	// QuoteComponent indicates the component of a quote to use when updating the indicator with a quote sample.
	//
	// If zero, the default (QuoteMidPrice) is used and the component is not shown in the indicator mnemonic.
	QuoteComponent entities.QuoteComponent

	// TradeComponent indicates the component of a trade to use when updating the indicator with a trade sample.
	//
	// If zero, the default (TradePrice) is used and the component is not shown in the indicator mnemonic.
	TradeComponent entities.TradeComponent
}

// DefaultSmoothingFactorParams returns a [SmoothingFactorParams] value populated with Ehlers defaults.
func DefaultSmoothingFactorParams() *SmoothingFactorParams {
	return &SmoothingFactorParams{
		SmoothingFactor: 0.07,
		SignalLag:       9,
	}
}
